File: vi/codes/java/chapter_backtracking/preorder_traversal_i_compact.java
/**
 * File: preorder_traversal_i_compact.java
 * Created Time: 2023-04-16
 * Author: krahets (krahets@163.com)
 */

package chapter_backtracking;

import utils.*;
import java.util.*;

public class preorder_traversal_i_compact {
    static List<TreeNode> res;

    /* Duyệt trước (tiền tự): Ví dụ 1 */
    static void preOrder(TreeNode root) {
        if (root == null) {
            return;
        }
        if (root.val == 7) {
            // Ghi lại lời giải
            res.add(root);
        }
        preOrder(root.left);
        preOrder(root.right);
    }

    public static void main(String[] args) {
        TreeNode root = TreeNode.listToTree(Arrays.asList(1, 7, 3, 4, 5, 6, 7));
        System.out.println("\nKhởi tạo cây nhị phân");
        PrintUtil.printTree(root);

        // Duyệt trước
        res = new ArrayList<>();
        preOrder(root);

        System.out.println("\nXuất ra tất cả các nút có giá trị bằng 7");
        List<Integer> vals = new ArrayList<>();
        for (TreeNode node : res) {
            vals.add(node.val);
        }
        System.out.println(vals);
    }
}

File: vi/codes/javascript/chapter_backtracking/preorder_traversal_i_compact.js
/**
 * File: preorder_traversal_i_compact.js
 * Created Time: 2023-05-09
 * Author: Justin (xiefahit@gmail.com)
 */

const { arrToTree } = require('../modules/TreeNode');
const { printTree } = require('../modules/PrintUtil');

/* Duyệt trước: ví dụ 1 */
function preOrder(root, res) {
    if (root === null) {
        return;
    }
    if (root.val === 7) {
        // Ghi lại lời giải
        res.push(root);
    }
    preOrder(root.left, res);
    preOrder(root.right, res);
}

// Mã điều khiển
const root = arrToTree([1, 7, 3, 4, 5, 6, 7]);
console.log('\nKhởi tạo cây nhị phân');
printTree(root);

// Duyệt trước
const res = [];
preOrder(root, res);

console.log('\nXuất ra tất cả các nút có giá trị 7');
console.log(res.map((node) => node.val));
